### PR TITLE
Adding implicit semi-colon parser to Nemerle

### PR DESCRIPTION
**Pull request by antagonist112358**
Similarly to Scala, I thought it would be nice if Nemerle had an option
to use implicit semi-colon type syntax, where semicolons are optional.

To that end, I've added this feature to the trunk. It includes tests,
examples, etc. The logic is based on Token.NewLine so as to not break / change the existing parser too much.

Usage is either:
1) Add the "-isc" switch when building using ncc
2) Use "pragma implicit" (similarly to "pragma indent" as the first line
in a code file.

Note - I had some issues with VS2013, so some of the changes (not relating to the parser) have to do with that.
Please take a look at the "implicit-syntax.n" and "implicit-syntax2.n" files below for some more details.

Please let me know if you need more information or details.

Thanks!

Kevin D.

### CODE REVIEW

**Comment by antagonist112358** [comment]
Ok - Sorry, been busy working on other projects. And I also apologize for not contacting you guys first before I submitted the pull request.

Lets try this again:

I'd like to make nemerle more "scala-like" in the sense that semicolons are not required in most cases. To that end, I added the "implicit" mode, trying to mimic the paradigm you already had for "indent" mode.

Let me start by responding to your comments:
--------------------------------------------------------------------------------------------------------------------------
"Implicit absolutely not informative title."
    I totally agree - to that end, I'd like to remove "implicit" altogether. Since the pre-parser I wrote is just an extension of the existing "block mode" pre-parser, I would like to make the optional semicolon the default pre-parser. Since this has no effect on any current code written in nemerle (I've compiled the trunk in this configuration several times), it would be strictly additive (as in a new feature)

"'\u2029' | '\u0085' | '\u2028' - it is also a end of line symbol."
    Yup, you are right. I'll make this change in the next pull request.

"In the implisit mode we can't move part of a expression on the next line?"
    Yes, absolutely. Please look at the tests I wrote for the current implicit implementation. Basically I haven't found a case where it doesn't work as expected (certainly not claiming there isn't one, just that I haven't found one yet)

"In tests are mixed spaces with tabs."
    Unlike indent mode, like "block" mode, tabs and spaces are not significant. I can certainly use all consistent tabs or spaces in the next pull request, but this doesn't change the behavior.
--------------------------------------------------------------------------------------------------------------------------

The optional semicolon PreParser simply uses a ruleset to determine if the current line should be terminated in a semicolon (if it isn't already), and if it determines that it should, adds one to the token-set. It makes these determinations based on the last non-tab, non-space token in the current like, and sometimes the next non-tab, non-space token in the next line.

If you guys don't mind creating a branch for me to work on this addition, that would be awesome. To that end, if you guys also don't mind creating a few additional tests for the semicolon-less use cases, I'd be happy to add those into the regression tests.

**Comment by antagonist112358** [comment]
If the project collaborators are OK with the idea of this language feature, I'd like to retract this pull request, make the above described changes, add in any regression tests you think are necessary / important to validate the feature, and then resubmit the pull request.

Does this work for everyone?

File: ncc/parsing/PreParserImplicit.n [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+        def next = peek_token();
+        if (next is Token.NewLine) Some(next)
+        else None()
+    }
+    
+    /** Checks to see if the newline should be continued or can be interpreted as
+        an implicit semicolon.
+        Has to eat the newline character also, since this will screw-up the Parser.
+    */
+    check_ahead_newline(last_tok: Token, tok: Token) : void {
```
Comment (antagonist112358) [inline ncc/parsing/PreParserImplicit.n]: This is the main function which determines if a semi-colon should be added to the tokens output.

File: ncc/parsing/PreParserImplicit.n [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+        
+    }
+    
+    /** Handle standard situations when new bracket group is beginning
+        or there is erronous situation. Any non bracket token is
+        appended to current LooseGroup.
+
+        Throws PreParserException when there is unmatched end bracket.
+     */
+    protected override handle_default_token (current_begin : int, tok : Token, braces_cut_current = true, scan_ahead_for_replacements = true) : void
```
Comment (antagonist112358) [inline ncc/parsing/PreParserImplicit.n]: scan_ahead_for_replacements flag determines whether _handle_default_token_ should try to add an implicit semicolon at the end of a line.

File: ncc/parsing/PreParserImplicit.n [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+              }
+            }
+            Token.RoundGroup (loc + tok.Location, round_group, openBrace, closeBrace);
+
+          // finish current loose group
+          | Token.Comma =>
+            finish_current (current_begin, tok);
+            loop ()
+          | Token.NewLine => loop()
+          | _ => handle_default_token (current_begin, tok, false, false); loop ()
```
Comment (antagonist112358) [inline ncc/parsing/PreParserImplicit.n]: Never try to add a semicolon in a round group.

File: ncc/parsing/PreParserImplicit.n [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+            def group = finish_parent (parent_begin, current_begin);
+            def groupLoc = loc + tok.Location;
+            when (group != null)
+              group.Location = groupLoc;
+            Token.SquareGroup (groupLoc, group, openBrace, closeBrace);
+
+          // finish current loose group
+          | Token.Comma => finish_current (current_begin, tok); loop ()
+          | Token.NewLine => loop()
+          | _ => handle_default_token (current_begin, tok, false, false); loop ()
```
Comment (antagonist112358) [inline ncc/parsing/PreParserImplicit.n]: Never try to add a semicolon in a square group.

File: ncc/parsing/PreParserImplicit.n [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+        match (tok) {
+          // finish entire brace group
+          | Token.EndQuote as closeBrace =>
+            def group = finish_parent (parent_begin, current_begin);
+            Token.QuoteGroup (loc + tok.Location, group, openBrace, closeBrace);
+
+          // finish current loose group
+          | Token.Semicolon => finish_current (current_begin, tok); loop ()
+          | Token.NewLine => loop()
+          | _ => handle_default_token (current_begin, tok); loop ()
```
Comment (antagonist112358) [inline ncc/parsing/PreParserImplicit.n]: <b>This should be changed</b> - Should never try to add a semicolon in a quote group either.

**Comment by NN---** [comment]
Can you call it more explicitly like: #pragma implicit_semicolon  ?
Does it work well with syntax macros ?
What if I write

```
mysyntaxmacro
  x;
```

Won't it be treated as : 


```
mysyntaxmacro;
  x;
```

**Comment by antagonist112358** [comment]
In the latest rev, I dropped the pragma altogether, with the goal being to make implicit semicolon the default parser.

So yeah, macros are a problem.

To address this, the idea is to retry any failed macros in the form of:

some-macro(...);
x;

As if they were entered as:

some-macro(...) x;

This is still a problem for macros with syntax:

{keyword} - Token.Identifier
{expression} - PExpr

Like:

requires
param1 != null
{
...

I'm still trying to figure out a solution. but I'm open to ideas / suggestions. I've got the current implementation down to a few edge cases, but I don't know if I've covered all cases.

The current goal is to get the entire Nemerle trunk to compile with the new parser. Are there other macro cases that don't appear in the codebase I should consider?

**Comment by NN---** [comment]
You have all the snippets folder to compile.
You can try compiling other Nemerle libraries like: 
NoR: https://github.com/Ziaw/nor
Statechart: https://github.com/CodingUnit/Nemerle.Statechart
NemerleWeb: https://github.com/NemerleWeb/NemerleWeb/

I don't think it is a good idea for implicit semicolon cause it breaks syntax macros and if you make some heuristic with double check you slow down the compiler :(

**Comment by antagonist112358** [comment]
>I don't think it is a good idea for implicit semicolon cause it breaks syntax macros

Actually I think I can do you one better - the latest rev (which I haven't pushed into my repo to pull into this one yet) is specifically targeted to handle macros (with syntax)

> if you make some heuristic with double check you slow down the compiler

Understand the concern - I'm up to a few match statements (which I think expand to 4 or less if/else statements, worse case), plus a single "if" in the MainParser.

The logic is (to be blunt) stupid simple - because semicolon gets converted to end-of-group, you get: when trying to parse and expression (in MainParser::parse_expr(...)), and we hit an end-of-group, jump to sibling and loop.

I need to fix the PreParseImplicit class (should be simple, and I might have time tonight or tomorrow), but this iteration has no problem with macros, even with arbitrary nesting.

In other words: the example case from before, expanded to be nastier:
```
foreach(item in group)
  if (item.someCondition)
    doSomeAction(item)
  else
    doSomeOtherAction(item)
```

parses correctly.

**Comment by antagonist112358** [comment]
Macros are not allowed to have syntax that would otherwise break the PreParser, like perhaps:
```
macro uppercut(...)
syntax("{", "{", PExpr, ".", "[", "[")
```
correct?

**Comment by NN---** [comment]
Btw, are you familiar with lexical macros ?
Are you sure your changes work with them correctly ?.

For instance:
```nemerle
 macro lexer (group : Token) syntax ("lexer", group) { def str = group.ToString(); // это тоже чистый fake. <[ $(str : string) ]> }
```
